src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,413 @@
+ * graph. It uses Floyd Warshall algorithm for obtaining the transitivity graph.
+ */
+static void
+CreateTransitivityMatrix()
+{
+	HASH_SEQ_STATUS status;
+	FRelNode *frelNode = NULL;
+	int sourceCounter = 0;
+	int interVertexCounter = 0;
+	int destinationCounter = 0;
+	int curIndex = 0;
+
+	/* first, allocate memory for transitivity matrix */
+	frelGraph->transitivityMatrix = (bool **) palloc(frelGraph->nodeCount *
+													 sizeof(bool *));
+	while (curIndex < frelGraph->nodeCount)
+	{
+		frelGraph->transitivityMatrix[curIndex] = (bool *) palloc(frelGraph->nodeCount *
+																  sizeof(bool));
+		memset(frelGraph->transitivityMatrix[curIndex], false, frelGraph->nodeCount *
+			   sizeof(bool));
+		curIndex += 1;
+	}
+
+	hash_seq_init(&status, frelGraph->nodeMap);
+
+	/*
+	 * Create the initial adjacency matrix, that will be transformed to
+	 * transitivity matrix dynamically.
+	 */
+	while ((frelNode = (FRelNode *) hash_seq_search(&status)) != 0)
+	{
+		uint32 relationIndex = frelNode->index;
+		ListCell *nodeCell = NULL;
+		List *curAdjacencyList = frelNode->adjacencyList;
+
+
+		foreach(nodeCell, curAdjacencyList)
+		{
+			FRelNode *referencingFrelNode = (FRelNode *) lfirst(nodeCell);
+			uint32 referencingRelationIndex = referencingFrelNode->index;
+			frelGraph->transitivityMatrix[relationIndex][referencingRelationIndex] = true;
+		}
+	}
+
+	/*
+	 * Add all vertices one by one to the set of intermediate vertices.
+	 *
+	 * Before start of a iteration, we have transitivity values for all
+	 * pairs of vertices such that the transitivity values consider only
+	 * the vertices in set {0, 1, 2, .. k-1} as intermediate vertices.
+	 *
+	 * After the end of a iteration, vertex no. k is added to the set of
+	 * intermediate vertices and the set becomes {0, 1, .. k}
+	 */
+	for (interVertexCounter = 0; interVertexCounter < frelGraph->nodeCount;
+		 interVertexCounter++)
+	{
+		/* pick all vertices as source one by one */
+		for (sourceCounter = 0; sourceCounter < frelGraph->nodeCount; sourceCounter++)
+		{
+			/* pick all vertices as destination for the above picked source */
+			for (destinationCounter = 0; destinationCounter < frelGraph->nodeCount;
+				 destinationCounter++)
+			{
+				/*
+				 * If vertex k is on a path from i to j, then make sure that
+				 * the value of transitivityMatrix[i][j] is 1.
+				 */
+				frelGraph->transitivityMatrix[sourceCounter][destinationCounter] =
+					frelGraph->transitivityMatrix[sourceCounter][destinationCounter] ||
+					(frelGraph->transitivityMatrix[sourceCounter][interVertexCounter] &&
+					 frelGraph->transitivityMatrix[interVertexCounter][destinationCounter]);
+			}
+		}
+	}
+}
+
+
+/*
+ * GetForeignKeyRelation returns the list of oids affected or affecting given
+ * relation id. It uses the transitivity matrix which is computed once in a
+ * session.
+ */
+List *
+GetForeignKeyRelation(Oid relationId, bool isAffecting)
```

Comment:
I'm writing some code on top of this PR for adding this information to metadata cache and invalidate the graph. I found this API slightly confusing for the following reasons:

1) Instead of passing a boolean flag, we should have two separate functions with clear names. 

2) `GetForeignKeyRelation()` isn't really a descriptive name. As mentioned in (1) we should probably have the following APIs `ReferencingRelationIdList()` and `ReferencedRelationIdList()`